api/api_public_transaction_pool.go

```diff
@@ -552,3 +552,40 @@ func resend(s *PublicTransactionPoolAPI, ctx context.Context, sendArgs NewTxArgs
 
 	return common.Hash{}, fmt.Errorf("Transaction %#x not found", matchTx.Hash())
 }
+
+// This API doesn't allow rueqesters to use block hash according to ValidateSender function's parameter
+func (s *PublicTransactionPoolAPI) RecoverTransaction(ctx context.Context, encodedTx hexutil.Bytes, blockNumber rpc.BlockNumber) (common.Address, error) {
+	tx := new(types.Transaction)
+	if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
+		return common.Address{}, err
+	}
+
+	signer := types.MakeSigner(s.b.ChainConfig(), s.b.CurrentBlock().Number())
+	state, _, err := s.b.StateAndHeaderByNumber(ctx, blockNumber)
+	if err != nil {
+		return common.Address{}, err
+	}
+	_, err = tx.ValidateSender(signer, state, blockNumber.Uint64())
+	if err != nil {
+		return common.Address{}, err
+	}
+	return signer.Sender(tx)
+}
+
+func (s *PublicTransactionPoolAPI) RecoverMessage(
+	ctx context.Context, address common.Address, data, sig hexutil.Bytes, blockNumber rpc.BlockNumber,
+) (common.Address, error) {
+	pubkey, err := klayEthEcRecover(data, sig)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	state, _, err := s.b.StateAndHeaderByNumber(ctx, blockNumber)
+	if err != nil {
+		return common.Address{}, err
+	}
+	if state.GetKey(address).IsContainedKey(pubkey) {
```

Comment:
Related to my comment at AccountKeyLegacy.IsContaindKey(),
```suggestion
	if state.GetKey(address).Type() == AccountKeyTypeLegacy || state.GetKey(address).IsContainedKey(pubkey)  {
```

Reply:
Okay, I got your point.
Returning false in the legacy account is also weird since correct pubkey also return false.
Anyway it looks better to denote explicitly in the `recoverMessage`.

Reply:
I see. Returning false is also weird. I came up with an unambiguous method.
We could add `ValidateMember()` alongside `Validate()` which are very similar, but ValidateMember is satisfied with just one public key.

```go
type AccountKey interface {
	Validate(currentBlockNumber uint64, r RoleType, recoveredKeys []*ecdsa.PublicKey, from common.Address) bool
	ValidateMember(blockNum uint64, recoveredKey *ecdsa.PublicKey, from common.Address) bool
}
```

Reply:
Reflected [002ffc5](https://github.com/klaytn/klaytn/pull/1911/commits/002ffc540b33cf96e5ae0981aa28f37c4e43f556)
This review seems to include all your points.

Reply:
Thanks! That looks nice.